### PR TITLE
feat: storage node enters recovery when lagging behind during operation

### PR DESCRIPTION
**Pull request by halfprice**
## Description

Before this change, storage node can only enter recovery mode upon restarts.
However, a storage node can become lagging behind and slow processing events and eventually falling out of
the current committees' scope (2 epoch behind or more).

In this case, the node should directly enter recovery mode.

Resolve WAL-479

## Test plan

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.
For each box you select, include information after the relevant heading that describes the impact of your changes that
a user might notice and any actions they must take to implement updates. (Add release notes after the colon for each item)

- [ ] Storage node:
- [ ] Aggregator:
- [ ] Publisher:
- [ ] CLI:

### CODE REVIEW

**Comment by halfprice** [comment]
Hi @mlegner, this PR is ready for review~

File: crates/walrus-service/src/node/committee/committee_service.rs [hunk 1]
```diff
@@ -574,35 +575,63 @@ where
...
-        } else if new_epoch < expected_next_epoch - 1 {
-            return Err(BeginCommitteeChangeError::EpochIsLess {
-                expected: expected_next_epoch,
-                actual: new_epoch,
-            });
         }
-        debug_assert_eq!(new_epoch, expected_next_epoch);
 
+        // Now, fetch the latest committee from the chain.
         let latest = self
```
Comment (mlegner) [inline crates/walrus-service/src/node/committee/committee_service.rs]: Nit (pre-existing): Can we call this `latest_committees` or something like that?

File: crates/walrus-service/src/node.rs [hunk 2]
```diff
@@ -1384,9 +1386,50 @@ impl StorageNode {
         Ok(())
     }
 
+    /// Storage node execution of the epoch change start event.
+    async fn execute_epoch_change_start_event(
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit/question: It's a bit unclear what the relation is between "process" and "execute".

File: crates/walrus-service/src/node.rs [hunk 3]
```diff
@@ -1600,26 +1639,45 @@ impl StorageNode {
...
+                debug_assert!(requested_epoch < latest_epoch);
+                // We are processing a backlog of events. Since the committee service has a
+                // more recent committee. In this situation, we have already lost the information
+                // and the shard assignment of the previous epoch relative to `event.epoch`, the
+                // node cannot execute the epoch change. Therefore, the node needs to enter recovery
+                // mode to catch up to the latest epoch as quickly as possible.
+                tracing::warn!(
+                    ?latest_epoch,
+                    ?requested_epoch,
+                    "epoch change requested for an older epoch than the latest epoch, this \
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs]: Nit:
```suggestion
                    "epoch change requested for an older epoch than the latest epoch; this means \
```

File: crates/walrus-service/src/node.rs [hunk 4]
```diff
@@ -1579,15 +1617,16 @@ impl StorageNode {
...
-    /// Initiates a committee transition to a new epoch.
+    /// Initiates a committee transition to a new epoch. Upon the return of this function, the
+    /// latest committee on chain is updated to the new node.
     ///
-    /// Returns `true` if epoch change event has started or was sufficiently recent such
-    /// that it should be handled.
+    /// Returns the action to execute epoch change based on the result of committee service,
+    /// including possible actions to enter recovery mode due to the node being severely lagging.
     #[tracing::instrument(skip_all)]
     async fn begin_committee_change(
```
Comment (mlegner) [inline crates/walrus-service/src/node.rs:1629]: Remark: AFAICT, the `CommitteeService::begin_committee_change` function is only used here (+ some tests). This results in effectively doing (almost) the same case distinction twice. Maybe this could be simplified (in a separate refactoring)?

File: crates/walrus-simtest/tests/simtest_failure.rs [hunk 5]
```diff
@@ -557,4 +561,160 @@ mod tests {
...
+                    .find(|s| s.shard == shard.shard);
+                assert!(
+                    shard_i_status.is_none()
+                        || shard_i_status.unwrap().status != ShardStatus::ReadOnly
+                );
+            }
+        }
+
+        assert_eq!(
+            simtest_utils::get_nodes_health_info(&[&walrus_cluster.nodes[0]])
```
Comment (mlegner) [inline crates/walrus-simtest/tests/simtest_failure.rs]: Question: Why do we have to get the health info again instead of reusing the `node_health_info`?

File: crates/walrus-simtest/tests/simtest_failure.rs [hunk 6]
```diff
@@ -557,4 +561,160 @@ mod tests {
...
+            .build_generic::<SimStorageNodeHandle>()
+            .await
+            .unwrap();
+
+        let blob_info_consistency_check = BlobInfoConsistencyCheck::new();
+
+        let client_arc = Arc::new(client);
+
+        // Starts a background workload that a client keeps writing and retrieving data.
+        // All requests should succeed even if a node crashes.
```
Comment (mlegner) [inline crates/walrus-simtest/tests/simtest_failure.rs]: Nit: There are several mentions of "crash" in this test. AFAICT, the node doesn't actually crash but just lag. Can you change some of the comments to avoid the confusion?

**Review by mlegner — APPROVED** [review]
Thanks a lot for taking care of this corner case! 🙏

File: crates/walrus-service/src/node.rs (hunk 3 above)
Comment (halfprice) [reply crates/walrus-service/src/node.rs]: Done

File: crates/walrus-simtest/tests/simtest_failure.rs (hunk 5 above)
Comment (halfprice) [reply crates/walrus-simtest/tests/simtest_failure.rs]: Good catch!

File: crates/walrus-simtest/tests/simtest_failure.rs (hunk 6 above)
Comment (halfprice) [reply crates/walrus-simtest/tests/simtest_failure.rs]: Done.

File: crates/walrus-service/src/node/committee/committee_service.rs (hunk 1 above)
Comment (halfprice) [reply crates/walrus-service/src/node/committee/committee_service.rs]: Done.

File: crates/walrus-service/src/node.rs (hunk 2 above)
Comment (halfprice) [reply crates/walrus-service/src/node.rs]: what my original thoughts are:

process means process epoch change event, which includes everything the node needs to do upon receiving epoch change start event, including things like consistency check that is not really part of the epoch change execution logic.

execute means the node actually executes the epoch change logic to bring the node state to the next epoch.

So I replaced `execute_epoch_change_start` to `execute_epoch_change`, and added some more documentation to make it clear.

File: crates/walrus-service/src/node.rs (hunk 4 above)
Comment (halfprice) [reply crates/walrus-service/src/node.rs:1629]: Good suggestion. Will tackle it in a follow up PR.
